// @dendronhq/api-server/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessReturnType = exports.ServerUtils = exports.SubProcessExitType = exports.getWSEngine = exports.putWS = exports.getWSKey = void 0;
const common_all_1 = require("@dendronhq/common-all");
const common_server_1 = require("@dendronhq/common-server");
const execa_1 = __importDefault(require("execa"));
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const _1 = require(".");
const memoryStore_1 = require("./store/memoryStore");
function getWSKey(uri) {
    return lodash_1.default.trimEnd(uri, "/").toLowerCase();
}
exports.getWSKey = getWSKey;
async function putWS({ ws, engine, }) {
    memoryStore_1.MemoryStore.instance().put(`ws:${getWSKey(ws)}`, engine);
}
exports.putWS = putWS;
async function getWSEngine({ ws }) {
    const engine = await memoryStore_1.MemoryStore.instance().get(`ws:${getWSKey(ws)}`);
    if (!engine) {
        throw `No Engine: ${ws}`;
    }
    return engine;
}
exports.getWSEngine = getWSEngine;
var SubProcessExitType;
(function (SubProcessExitType) {
    SubProcessExitType["EXIT"] = "exit";
    SubProcessExitType["SIGINT"] = "SIGINT";
    SubProcessExitType["SIGURS1"] = "SIGUSR1";
    SubProcessExitType["SIGURS2"] = "SIGUSR2";
    SubProcessExitType["UNCAUGHT_EXCEPTION"] = "uncaughtException";
})(SubProcessExitType = exports.SubProcessExitType || (exports.SubProcessExitType = {}));
class ServerUtils {
    static onProcessExit({ subprocess, cb, }) {
        subprocess.on("exit", () => cb(SubProcessExitType.EXIT));
        subprocess.on("SIGINT", () => cb(SubProcessExitType.SIGINT));
        // catches "kill pid" (for example: nodemon restart)
        subprocess.on("SIGUSR1", () => cb(SubProcessExitType.SIGURS1));
        subprocess.on("SIGUSR2", () => cb(SubProcessExitType.SIGURS2));
        //catches uncaught exceptions
        subprocess.on("uncaughtException", () => cb(SubProcessExitType.UNCAUGHT_EXCEPTION));
    }
    /**
     * Attach to a server process to kill it when the current process exits
     * @param subprocess
     */
    static cleanServerProcess(subprocess) {
        const handleExit = () => {
            console.log("handle exit");
            try {
                process.kill(subprocess.pid);
            }
            catch (err) {
                // this means process was already killed
                if (err.code !== "ESRCH") {
                    console.log("process already killed");
                    throw err;
                }
            }
        };
        ServerUtils.onProcessExit({ subprocess, cb: handleExit });
    }
    static prepareServerArgs() {
        const { NEXT_SERVER_URL, NEXT_STATIC_ROOT, ENGINE_SERVER_PORT, LOG_PATH, GOOGLE_OAUTH_ID, GOOGLE_OAUTH_SECRET, } = process.env;
        if (lodash_1.default.some(["LOG_PATH"], (k) => {
            return lodash_1.default.isUndefined(process.env[k]);
        })) {
            throw new common_all_1.DendronError({
                message: "no value found for env variable",
                status: common_all_1.ERROR_STATUS.INVALID_CONFIG,
            });
        }
        const logPath = LOG_PATH;
        let port;
        if (ENGINE_SERVER_PORT) {
            port = parseInt(ENGINE_SERVER_PORT, 10);
        }
        const nextServerUrl = NEXT_SERVER_URL;
        const nextStaticRoot = NEXT_STATIC_ROOT;
        const googleOauthClientId = GOOGLE_OAUTH_ID;
        const googleOauthClientSecret = GOOGLE_OAUTH_SECRET;
        return {
            port,
            logPath,
            nextServerUrl,
            nextStaticRoot,
            googleOauthClientId,
            googleOauthClientSecret,
        };
    }
    /**
     * Launch engine server
     * @param
     * @returns
     */
    static async startServerNode({ logPath, nextServerUrl, nextStaticRoot, port, googleOauthClientId, googleOauthClientSecret, }) {
        const { port: finalPort } = await (0, _1.launchv2)({
            port,
            logPath: path_1.default.join(logPath, "dendron.server.log"),
            nextServerUrl,
            nextStaticRoot,
            googleOauthClientId,
            googleOauthClientSecret,
        });
        if (!process.send) {
            throw new common_all_1.DendronError({ message: "expect a child process" });
        }
        process.send(`${finalPort}`);
        return { port: finalPort };
    }
    /**
     * Create a subprocess with a running instance of the engine server
     * @returns
     */
    static async execServerNode({ scriptPath, logPath, nextServerUrl, nextStaticRoot, port, googleOauthClientId, googleOauthClientSecret, }) {
        const logger = (0, common_server_1.createLogger)("execServer", path_1.default.join(logPath, "dendron.log"));
        return new Promise((resolve, reject) => {
            logger.info({ state: "enter" });
            const subprocess = execa_1.default.node(scriptPath, {
                env: {
                    LOG_PATH: logPath,
                    ENGINE_SERVER_PORT: port,
                    NEXT_SERVER_URL: nextServerUrl,
                    NEXT_STATIC_ROOT: nextStaticRoot,
                    GOOGLE_OAUTH_ID: googleOauthClientId,
                    GOOGLE_OAUTH_SECRET: googleOauthClientSecret,
                    ELECTRON_RUN_AS_NODE: 1,
                },
            });
            logger.info({ state: "post:exec.node" });
            subprocess.on("close", (code) => {
                logger.error({ state: "close" });
                reject(new common_all_1.DendronError({ message: "close", payload: { code } }));
            });
            subprocess.on("disconnect", () => {
                logger.error({ state: "disconnect" });
                reject(new common_all_1.DendronError({ message: "disconnect" }));
            });
            subprocess.on("exit", (code) => {
                logger.error({ state: "exit" });
                reject(new common_all_1.DendronError({ message: "exit", payload: { code } }));
            });
            subprocess.on("error", (err) => {
                logger.error({ state: "error", payload: err });
                reject(new common_all_1.DendronError({ message: "error", payload: (0, common_all_1.stringifyError)(err) }));
            });
            subprocess.on("message", (message) => {
                logger.info({ state: "message", message });
                const port = parseInt(message, 10);
                if (port <= 0) {
                    reject({
                        error: new common_all_1.DendronError({
                            message: "port is smaller than 0",
                            severity: common_all_1.ERROR_SEVERITY.FATAL,
                        }),
                    });
                }
                resolve({ port, subprocess });
            });
            this.cleanServerProcess(subprocess);
        });
    }
}
exports.ServerUtils = ServerUtils;
var ProcessReturnType;
(function (ProcessReturnType) {
    ProcessReturnType["ERROR"] = "error";
})(ProcessReturnType = exports.ProcessReturnType || (exports.ProcessReturnType = {}));
