// @dendronhq/common-all/lib/util/compat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.offsetRange = exports.newRange = exports.getTextRange = exports.position2VSCodeRange = exports.point2VSCodePosition = void 0;
/** Convert a `Point` from a parsed remark node to a `vscode.Poisition`
 *
 * @param point The point to convert.
 * @param offset When converting the point, shift it by this much.
 * @returns The converted Position, shifted by `offset` if provided.
 */
function point2VSCodePosition(point, offset) {
    return {
        // remark Point's are 0 indexed
        line: point.line - 1 + ((offset === null || offset === void 0 ? void 0 : offset.line) || 0),
        character: point.column - 1 + ((offset === null || offset === void 0 ? void 0 : offset.column) || 0),
    };
}
exports.point2VSCodePosition = point2VSCodePosition;
/** Convert a `Position` from a parsed remark node to a `vscode.Range`
 *
 * @param position The position to convert.
 * @returns The converted Range.
 */
function position2VSCodeRange(position, offset) {
    return {
        // remark Point's are 0 indexed
        start: point2VSCodePosition(position.start, offset),
        end: point2VSCodePosition(position.end, offset),
    };
}
exports.position2VSCodeRange = position2VSCodeRange;
/** Similar to VSCode's `Document.getRange`, except that it works with strings. */
function getTextRange(text, range) {
    const { start, end } = range;
    const lines = text.split("\n").splice(start.line, end.line - start.line + 1);
    if (lines.length === 0)
        return "";
    // Do the end first in case there is just one line, otherwise the end of the string would shift
    lines[lines.length - 1] = lines[lines.length - 1].substring(0, end.character);
    lines[0] = lines[0].substring(start.character);
    return lines.join("\n");
}
exports.getTextRange = getTextRange;
function newRange(startLine, startCharacter, endLine, endCharacter) {
    return {
        start: {
            line: startLine,
            character: startCharacter,
        },
        end: {
            line: endLine,
            character: endCharacter,
        },
    };
}
exports.newRange = newRange;
function offsetRange(range, offset) {
    const { line, column } = { line: 0, column: 0, ...offset };
    return newRange(range.start.line + line, range.start.character + column, range.end.line + line, range.end.character + column);
}
exports.offsetRange = offsetRange;
