// @dendronhq/pods-core/lib/v2/ConfigFileUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigFileUtils = void 0;
const common_all_1 = require("@dendronhq/common-all");
const common_server_1 = require("@dendronhq/common-server");
const fs_extra_1 = __importStar(require("fs-extra"));
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const __1 = require("..");
class ConfigFileUtils {
    static getConfigByFPath({ fPath }) {
        if (!fs_extra_1.default.existsSync(fPath)) {
            return undefined;
        }
        else {
            return (0, common_server_1.readYAML)(fPath);
        }
    }
    /**
     * Create config file if it doesn't exist
     */
    static genConfigFileV2({ fPath, configSchema, force, setProperties, }) {
        (0, fs_extra_1.ensureDirSync)(path_1.default.dirname(fPath));
        const required = configSchema.required;
        const podConfig = configSchema.properties;
        const config = Object.keys(podConfig)
            .map((ent) => {
            podConfig[ent] = lodash_1.default.defaults(podConfig[ent], { default: "TODO" });
            const args = [
                `# description: ${podConfig[ent].description}`,
                `# type: ${podConfig[ent].type}`,
            ];
            let configPrefix = "# ";
            if (required.includes(`${ent}`)) {
                args.push(`# required: true`);
                configPrefix = "";
            }
            // If a pre-set value for an optional param has been passed in, then keep it.
            if (setProperties && ent in setProperties) {
                configPrefix = "";
            }
            /**
             * add NOTE for config options having note property. Mark the option as a comment.
             * added for exportScope config option
             */
            if (podConfig[ent].note) {
                args.push(`# NOTE: ${podConfig[ent].note}`);
                configPrefix = "# ";
            }
            args.push(`${configPrefix}${ent}: ${setProperties && ent in setProperties
                ? setProperties[ent]
                : podConfig[ent].default}`);
            return args.join("\n");
        })
            .join("\n\n");
        const fileExists = fs_extra_1.default.existsSync(fPath);
        if (!fileExists || force) {
            (0, fs_extra_1.writeFileSync)(fPath, config);
        }
        else if (fileExists && !force) {
            throw new Error("Config already exists!");
        }
        return fPath;
    }
    static createExportConfig(opts) {
        return {
            type: "object",
            required: ["podId", "podType", ...opts.required],
            properties: {
                podId: {
                    description: "configuration ID",
                    type: "string",
                },
                description: {
                    description: "optional description for the pod",
                    type: "string",
                    nullable: true,
                },
                exportScope: {
                    description: "export scope of the pod.",
                    type: "string",
                    note: "When a setting is missing from this config, you will get a UI prompt to select a value for that setting while running the export pod. For this particular exportScope setting, if you would rather not be prompted and always have the same exportScope, simply uncomment the line below.",
                },
                podType: {
                    description: "type of pod",
                    type: "string",
                },
                ...opts.properties,
            },
        };
    }
    static getConfigSchema(podType) {
        switch (podType) {
            case __1.PodV2Types.AirtableExportV2:
                return __1.AirtableExportPodV2.config();
            case __1.PodV2Types.GoogleDocsExportV2:
                return __1.GoogleDocsExportPodV2.config();
            case __1.PodV2Types.MarkdownExportV2:
                return __1.MarkdownExportPodV2.config();
            case __1.PodV2Types.NotionExportV2:
                return __1.NotionExportPodV2.config();
            case __1.PodV2Types.JSONExportV2:
                return __1.JSONExportPodV2.config();
            default:
                (0, common_all_1.assertUnreachable)(podType);
        }
    }
}
exports.ConfigFileUtils = ConfigFileUtils;
