// @dendronhq/common-all/lib/types/configs/workspace/task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskNoteUtils = exports.genDefaultTaskConfig = void 0;
const __1 = require("../..");
const types_1 = require("./types");
/**
 * Generates default {@link ScratchConfig}
 * @returns ScratchConfig
 */
function genDefaultTaskConfig() {
    return {
        name: "task",
        dateFormat: "y.MM.dd",
        addBehavior: types_1.NoteAddBehaviorEnum.asOwnDomain,
        statusSymbols: {
            "": " ",
            wip: "w",
            done: "x",
            assigned: "a",
            moved: "m",
            blocked: "b",
            delegated: "l",
            dropped: "d",
            pending: "y",
        },
        taskCompleteStatus: ["done", "x"],
        prioritySymbols: {
            H: "high",
            M: "medium",
            L: "low",
        },
        todoIntegration: false,
        createTaskSelectionType: __1.LegacyLookupSelectionType.selection2link,
    };
}
exports.genDefaultTaskConfig = genDefaultTaskConfig;
/** Used to recognize if something is a task note. If any of these are included in the frontmatter, the note will be considered a task note. */
const TASK_NOTE_PROP_KEYS = ["status", "due", "owner", "priority"];
class TaskNoteUtils {
    static isTaskNote(note) {
        for (const prop of TASK_NOTE_PROP_KEYS) {
            if (note.custom !== undefined && note.custom[prop] !== undefined)
                return true;
        }
        return false;
    }
    static genDefaultTaskNoteProps(note, config) {
        const props = {
            custom: {
                status: "",
                due: "",
                priority: "",
                owner: "",
            },
        };
        if (config.todoIntegration)
            props.custom["TODO"] = note.title;
        return props;
    }
    static getStatusSymbolRaw({ note, taskConfig, }) {
        const { status } = note.custom;
        if (status === undefined)
            return undefined;
        // If the symbol is not mapped to anything, use the symbol prop directly
        if (!taskConfig.statusSymbols)
            return `${status}`;
        const symbol = taskConfig.statusSymbols[status];
        if (symbol === undefined)
            return `${status}`;
        // If it does map to something, then use that
        return `${symbol}`;
    }
    static getStatusSymbol(props) {
        const status = this.getStatusSymbolRaw(props);
        if (status === undefined)
            return undefined;
        return `[${this.getStatusSymbolRaw(props)}]`;
    }
    static isTaskComplete({ note, taskConfig, }) {
        var _a;
        const { status } = note.custom;
        return status && ((_a = taskConfig.taskCompleteStatus) === null || _a === void 0 ? void 0 : _a.includes(status));
    }
    static getPrioritySymbol({ note, taskConfig, }) {
        const { priority } = note.custom;
        if (priority === undefined)
            return undefined;
        // If the symbol is not mapped to anything, use the symbol prop directly
        if (!taskConfig.prioritySymbols)
            return priority;
        const symbol = taskConfig.prioritySymbols[priority];
        if (symbol === undefined)
            return priority;
        // If it does map to something, then use that
        return symbol;
    }
}
exports.TaskNoteUtils = TaskNoteUtils;
