// @dendronhq/common-all/lib/types/configs/workspace/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genDefaultWorkspaceConfig = void 0;
const journal_1 = require("./journal");
const scratch_1 = require("./scratch");
const graph_1 = require("../workspace/graph");
const base_1 = require("../base");
const task_1 = require("./task");
/**
 * Generates default {@link DendronWorkspaceConfig}
 * @returns DendronWorkspaceConfig
 */
function genDefaultWorkspaceConfig() {
    return {
        vaults: [],
        journal: (0, journal_1.genDefaultJournalConfig)(),
        scratch: (0, scratch_1.genDefaultScratchConfig)(),
        task: (0, task_1.genDefaultTaskConfig)(),
        graph: (0, graph_1.genDefaultGraphConfig)(),
        enableAutoCreateOnDefinition: false,
        enableXVaultWikiLink: false,
        enableRemoteVaultInit: true,
        enableUserTags: true,
        enableHashTags: true,
        workspaceVaultSyncMode: base_1.VaultSyncModeEnum.noCommit,
        enableAutoFoldFrontmatter: false,
        enableEditorDecorations: true,
        maxPreviewsCached: 10,
        maxNoteLength: 204800,
        enableFullHierarchyNoteTitle: false,
        enableSmartRefs: false,
    };
}
exports.genDefaultWorkspaceConfig = genDefaultWorkspaceConfig;
