// @dendronhq/engine-server/lib/metadata/service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataService = exports.WorkspaceActivationContext = exports.LapsedUserSurveyStatusEnum = exports.InitialSurveyStatusEnum = exports.InactvieUserMsgStatusEnum = exports.PriorTools = exports.ShowcaseEntry = void 0;
const common_all_1 = require("@dendronhq/common-all");
const fs_extra_1 = __importDefault(require("fs-extra"));
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
var ShowcaseEntry;
(function (ShowcaseEntry) {
    ShowcaseEntry["TryMeetingNotes"] = "TryMeetingNotes";
    ShowcaseEntry["AutocompleteTip"] = "AutocompleteTip";
    ShowcaseEntry["TagsTip"] = "TagsTip";
    ShowcaseEntry["RenameHeader"] = "RenameHeader";
    ShowcaseEntry["TaskManagement"] = "TaskManagement";
    ShowcaseEntry["BlockRefs"] = "BlockRefs";
    ShowcaseEntry["HeaderRefs"] = "HeaderRefs";
    ShowcaseEntry["InsertNoteLink"] = "InsertNoteLink";
    ShowcaseEntry["GraphTheme"] = "GraphTheme";
    ShowcaseEntry["PublishTheme"] = "PublishTheme";
    ShowcaseEntry["PreviewTheme"] = "PreviewTheme";
    ShowcaseEntry["GraphPanel"] = "GraphPanel";
    ShowcaseEntry["BacklinksPanelHover"] = "BacklinksPanelHover";
    ShowcaseEntry["ObsidianImport"] = "ObsidianImport";
})(ShowcaseEntry = exports.ShowcaseEntry || (exports.ShowcaseEntry = {}));
/**
 * Survey for users on which prior note-taking tools they've used.
 */
var PriorTools;
(function (PriorTools) {
    PriorTools["No"] = "No";
    PriorTools["Foam"] = "Foam";
    PriorTools["Roam"] = "Roam";
    PriorTools["Logseq"] = "Logseq";
    PriorTools["Notion"] = "Notion";
    PriorTools["OneNote"] = "OneNote";
    PriorTools["Obsidian"] = "Obsidian";
    PriorTools["Evernote"] = "Evernote";
    PriorTools["GoogleKeep"] = "Google Keep";
    PriorTools["Confluence"] = "Confluence";
    PriorTools["Other"] = "Other";
})(PriorTools = exports.PriorTools || (exports.PriorTools = {}));
var InactvieUserMsgStatusEnum;
(function (InactvieUserMsgStatusEnum) {
    InactvieUserMsgStatusEnum["submitted"] = "submitted";
    InactvieUserMsgStatusEnum["cancelled"] = "cancelled";
})(InactvieUserMsgStatusEnum = exports.InactvieUserMsgStatusEnum || (exports.InactvieUserMsgStatusEnum = {}));
var InitialSurveyStatusEnum;
(function (InitialSurveyStatusEnum) {
    InitialSurveyStatusEnum["submitted"] = "submitted";
    InitialSurveyStatusEnum["cancelled"] = "cancelled";
})(InitialSurveyStatusEnum = exports.InitialSurveyStatusEnum || (exports.InitialSurveyStatusEnum = {}));
var LapsedUserSurveyStatusEnum;
(function (LapsedUserSurveyStatusEnum) {
    LapsedUserSurveyStatusEnum["submitted"] = "submitted";
    LapsedUserSurveyStatusEnum["cancelled"] = "cancelled";
})(LapsedUserSurveyStatusEnum = exports.LapsedUserSurveyStatusEnum || (exports.LapsedUserSurveyStatusEnum = {}));
var WorkspaceActivationContext;
(function (WorkspaceActivationContext) {
    // UNSET - Indicates this is the first Workspace Launch
    WorkspaceActivationContext[WorkspaceActivationContext["normal"] = 0] = "normal";
    WorkspaceActivationContext[WorkspaceActivationContext["tutorial"] = 1] = "tutorial";
    WorkspaceActivationContext[WorkspaceActivationContext["seedBrowser"] = 2] = "seedBrowser";
})(WorkspaceActivationContext = exports.WorkspaceActivationContext || (exports.WorkspaceActivationContext = {}));
let _singleton;
class MetadataService {
    static instance() {
        if (!_singleton) {
            _singleton = new MetadataService();
        }
        return _singleton;
    }
    static metaFilePath() {
        return path_1.default.join(os_1.default.homedir(), common_all_1.FOLDERS.DENDRON_SYSTEM_ROOT, "meta.json");
    }
    deleteMeta(key) {
        const stateFromFile = this.getMeta();
        delete stateFromFile[key];
        fs_extra_1.default.writeJSONSync(MetadataService.metaFilePath(), stateFromFile, {
            spaces: 4,
        });
    }
    getMeta() {
        const metaPath = MetadataService.metaFilePath();
        if (!fs_extra_1.default.pathExistsSync(metaPath)) {
            fs_extra_1.default.ensureFileSync(metaPath);
            fs_extra_1.default.writeJSONSync(metaPath, {});
            return {};
        }
        return fs_extra_1.default.readJSONSync(MetadataService.metaFilePath());
    }
    getFeatureShowcaseStatus(key) {
        const featureShowcaseData = this.getMeta().featureShowcase;
        if (!featureShowcaseData) {
            return undefined;
        }
        return featureShowcaseData[key];
    }
    getGlobalVersion() {
        return this.getMeta().version || "0.0.0";
    }
    getLapsedUserSurveyStatus() {
        return this.getMeta().lapsedUserSurveyStatus;
    }
    getActivationContext() {
        var _a;
        return ((_a = this.getMeta().workspaceActivationContext) !== null && _a !== void 0 ? _a : WorkspaceActivationContext.normal);
    }
    get TipOfDayIndex() {
        return this.getMeta().tipOfTheDayIndex;
    }
    getGraphTheme() {
        return this.getMeta().graphTheme;
    }
    getTreeViewItemLabelType() {
        return (this.getMeta().treeViewItemLabelType || common_all_1.TreeViewItemLabelTypeEnum.title);
    }
    get BacklinksPanelSortOrder() {
        return this.getMeta().backlinksPanelSortOrder;
    }
    get priorTools() {
        return this.getMeta().priorTools;
    }
    get RecentWorkspaces() {
        return this.getMeta().recentWorkspaces;
    }
    get graphDepth() {
        return this.getMeta().graphDepth;
    }
    get graphPanelShowBacklinks() {
        return this.getMeta().graphPanelShowBacklinks;
    }
    get graphPanelShowOutwardLinks() {
        return this.getMeta().graphPanelShowOutwardLinks;
    }
    get graphPanelShowHierarchy() {
        return this.getMeta().graphPanelShowHierarchy;
    }
    setMeta(key, value) {
        const stateFromFile = this.getMeta();
        stateFromFile[key] = value;
        fs_extra_1.default.writeJSONSync(MetadataService.metaFilePath(), stateFromFile, {
            spaces: 4,
        });
    }
    get v100ReleaseMessageShown() {
        return this.getMeta().v100ReleaseMessageShown;
    }
    /**
     * Set first install logic
     *  ^o4y7ijuvi5nv
     */
    setInitialInstall(time) {
        time || (time = common_all_1.Time.now().toSeconds());
        this.setMeta("firstInstall", time);
    }
    setFirstWsInitialize() {
        this.setMeta("firstWsInitialize", common_all_1.Time.now().toSeconds());
    }
    setLapsedUserMsgSendTime() {
        this.setMeta("lapsedUserMsgSendTime", common_all_1.Time.now().toSeconds());
    }
    setDendronWorkspaceActivated() {
        this.setMeta("dendronWorkspaceActivated", common_all_1.Time.now().toSeconds());
    }
    setFirstLookupTime() {
        this.setMeta("firstLookupTime", common_all_1.Time.now().toSeconds());
    }
    setLastLookupTime() {
        this.setMeta("lastLookupTime", common_all_1.Time.now().toSeconds());
    }
    setInactiveUserMsgSendTime() {
        this.setMeta("inactiveUserMsgSendTime", common_all_1.Time.now().toSeconds());
    }
    setInactiveUserMsgStatus(value) {
        this.setMeta("inactiveUserMsgStatus", value);
    }
    setInitialSurveyStatus(value) {
        this.setMeta("initialSurveyStatus", value);
    }
    setLapsedUserSurveyStatus(value) {
        this.setMeta("lapsedUserSurveyStatus", value);
    }
    setGlobalVersion(value) {
        this.setMeta("version", value);
    }
    setFeatureShowcaseStatus(key) {
        const meta = this.getMeta();
        if (!meta.featureShowcase) {
            meta.featureShowcase = {};
        }
        meta.featureShowcase[key] = common_all_1.Time.now().toSeconds();
        return this.setMeta("featureShowcase", meta.featureShowcase);
    }
    setActivationContext(context) {
        this.setMeta("workspaceActivationContext", context);
    }
    set TipOfDayIndex(index) {
        this.setMeta("tipOfTheDayIndex", index);
    }
    setGraphTheme(graphTheme) {
        const meta = this.getMeta();
        if (meta.graphTheme !== graphTheme) {
            this.setMeta("graphTheme", graphTheme);
        }
    }
    set graphDepth(graphDepth) {
        const meta = this.getMeta();
        if (meta.graphDepth !== graphDepth) {
            this.setMeta("graphDepth", graphDepth);
        }
    }
    setTreeViewItemLabelType(labelType) {
        this.setMeta("treeViewItemLabelType", labelType);
    }
    set BacklinksPanelSortOrder(sortOrder) {
        this.setMeta("backlinksPanelSortOrder", sortOrder);
    }
    setFirstDailyJournalTime() {
        this.setMeta("firstDailyJournalTime", common_all_1.Time.now().toSeconds());
    }
    set priorTools(priorTools) {
        this.setMeta("priorTools", priorTools);
    }
    set v100ReleaseMessageShown(hasShown) {
        this.setMeta("v100ReleaseMessageShown", hasShown);
    }
    set graphPanelShowBacklinks(showBacklinks) {
        this.setMeta("graphPanelShowBacklinks", showBacklinks);
    }
    set graphPanelShowOutwardLinks(showOutwardLinks) {
        this.setMeta("graphPanelShowOutwardLinks", showOutwardLinks);
    }
    set graphPanelShowHierarchy(showHierarchy) {
        this.setMeta("graphPanelShowHierarchy", showHierarchy);
    }
    // Add a single path to recent workspaces. Recent workspaces acts like a FIFO
    // queue
    addToRecentWorkspaces(path) {
        const RECENT_WORKSPACE_ITEM_LIMIT = 5;
        const current = this.getMeta().recentWorkspaces;
        const updated = [];
        if (!current) {
            updated.push(path);
        }
        else {
            current.forEach((existingPath) => {
                if (existingPath !== path) {
                    updated.push(existingPath);
                }
            });
            if (updated.length >= RECENT_WORKSPACE_ITEM_LIMIT) {
                updated.pop();
            }
            // The first element of the array is the most recent
            updated.unshift(path);
        }
        this.setMeta("recentWorkspaces", updated);
    }
}
exports.MetadataService = MetadataService;
