// @dendronhq/unified/lib/decorations/decorations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runAllDecorators = void 0;
const common_all_1 = require("@dendronhq/common-all");
const userTags_1 = require("./userTags");
const types_1 = require("../types");
const hashTags_1 = require("./hashTags");
const references_1 = require("./references");
const wikilinks_1 = require("./wikilinks");
const frontmatter_1 = require("./frontmatter");
const blockAnchors_1 = require("./blockAnchors");
const utilsv5_1 = require("../utilsv5");
const diagnostics_1 = require("./diagnostics");
const lodash_1 = __importDefault(require("lodash"));
const unist_util_visit_1 = __importDefault(require("unist-util-visit"));
const __1 = require("..");
/** Dispatches the correct decorator based on the type of AST node. */
function runDecorator(opts) {
    const { node } = opts;
    switch (node.type) {
        case types_1.DendronASTTypes.BLOCK_ANCHOR:
            return (0, blockAnchors_1.decorateBlockAnchor)(opts);
        case types_1.DendronASTTypes.HASHTAG:
            return (0, hashTags_1.decorateHashTag)(opts);
        case types_1.DendronASTTypes.FRONTMATTER:
            return (0, frontmatter_1.decorateFrontmatter)(opts);
        case types_1.DendronASTTypes.USERTAG:
            return (0, userTags_1.decorateUserTag)(opts);
        case types_1.DendronASTTypes.WIKI_LINK:
            return (0, wikilinks_1.decorateWikilink)(opts);
        case types_1.DendronASTTypes.REF_LINK_V2: // fall-through deliberate
            return (0, references_1.decorateReference)(opts);
        default:
            return undefined;
    }
}
/** Get all decorations within the visible ranges for given note. */
async function runAllDecorators(opts) {
    const { note, ranges, engine } = opts;
    const allDecorations = [];
    const allDiagnostics = [];
    const allErrors = [];
    const proc = utilsv5_1.MDUtilsV5.procRemarkParse({
        mode: utilsv5_1.ProcMode.FULL,
        parseOnly: true,
    }, {
        dest: types_1.DendronASTDest.MD_DENDRON,
        engine,
        vault: note.vault,
        fname: note.fname,
    });
    for (const { range, text } of ranges) {
        if (text.length > common_all_1.ConfigUtils.getWorkspace(engine.config).maxNoteLength) {
            return {
                errors: [
                    new common_all_1.DendronError({
                        message: `Stopping decorations because visible range is too large. Unless you have a massive screen or really long lines of text, this may be a bug.`,
                        payload: {
                            maxNoteLength: common_all_1.ConfigUtils.getWorkspace(engine.config)
                                .maxNoteLength,
                            textLength: text.length,
                        },
                    }),
                ],
            };
        }
        const tree = proc.parse(text);
        // eslint-disable-next-line no-await-in-loop
        await __1.MdastUtils.visitAsync(tree, [], async (nodeIn) => {
            // This was parsed, it must have a position
            const node = nodeIn;
            // Need to update node position with the added offset from the range
            const decoratorOut = await runDecorator({
                ...opts,
                node,
                note,
                noteText: text,
            });
            if (decoratorOut) {
                const { decorations, errors } = decoratorOut;
                allDecorations.push(...decorations.map((decoration) => {
                    // Add the offset from the start of the range so these decorations match up in the original document
                    decoration.range = (0, common_all_1.offsetRange)(decoration.range, {
                        line: range.start.line,
                    });
                    return decoration;
                }));
                if (errors)
                    allErrors.push(...errors);
            }
        });
    }
    // Check for frontmatter diagnostics. Diagnostics always run on the whole note because they need to be active even when they are not visible.
    let frontmatter;
    const fullTree = proc.parse(opts.text);
    (0, unist_util_visit_1.default)(fullTree, ["yaml"], (node) => {
        frontmatter = node;
        return false; // stop iterating
    });
    if (lodash_1.default.isUndefined(frontmatter)) {
        allDiagnostics.push((0, diagnostics_1.warnMissingFrontmatter)());
    }
    else {
        const { diagnostics, errors } = (0, diagnostics_1.checkAndWarnBadFrontmatter)(note, frontmatter);
        allDiagnostics.push(...diagnostics);
        if (errors)
            allErrors.push(...errors);
    }
    return {
        allDecorations,
        allDiagnostics,
        allErrors,
    };
}
exports.runAllDecorators = runAllDecorators;
