// @dendronhq/unified/lib/decorations/frontmatter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decorateFrontmatter = void 0;
const common_all_1 = require("@dendronhq/common-all");
const hashTags_1 = require("./hashTags");
const utils_1 = require("./utils");
const lodash_1 = __importDefault(require("lodash"));
const yaml_1 = require("../yaml");
const decorateFrontmatter = async (opts) => {
    const { node: frontmatter, engine } = opts;
    const { value: contents, position } = frontmatter;
    // Decorate the timestamps
    const entries = contents.split("\n");
    const lineOffset = (0, common_all_1.point2VSCodePosition)(position.start).line +
        1; /* `---` line of frontmatter */
    const timestampDecorations = entries
        .map((entry, line) => {
        var _a;
        const match = common_all_1.NoteUtils.RE_FM_UPDATED_OR_CREATED.exec(entry);
        if (!lodash_1.default.isNull(match) && ((_a = match.groups) === null || _a === void 0 ? void 0 : _a.timestamp)) {
            const timestamp = lodash_1.default.toInteger(match.groups.timestamp);
            const decoration = {
                range: {
                    start: {
                        line: line + lineOffset,
                        character: match.groups.beforeTimestamp.length,
                    },
                    end: {
                        line: line + lineOffset,
                        character: match.groups.beforeTimestamp.length +
                            match.groups.timestamp.length,
                    },
                },
                timestamp,
                type: utils_1.DECORATION_TYPES.timestamp,
            };
            return decoration;
        }
        return undefined;
    })
        .filter(common_all_1.isNotUndefined);
    // Decorate the frontmatter tags
    const tags = (0, yaml_1.getFrontmatterTags)((0, yaml_1.parseFrontmatter)(contents));
    const tagDecorations = [];
    const errors = [];
    await Promise.all(tags.map(async (tag) => {
        const { errors, decorations } = await (0, hashTags_1.decorateTag)({
            fname: `${common_all_1.TAGS_HIERARCHY}${tag.value}`,
            position: tag.position,
            lineOffset,
            engine,
        });
        tagDecorations.push(...decorations);
        errors.push(...errors);
    }));
    const decorations = [
        ...tagDecorations,
        ...timestampDecorations,
    ];
    return {
        decorations,
        errors,
    };
};
exports.decorateFrontmatter = decorateFrontmatter;
