// @dendronhq/api-server/lib/modules/schemas/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaController = void 0;
const common_all_1 = require("@dendronhq/common-all");
const utils_1 = require("../../utils");
class SchemaController {
    static instance() {
        if (!SchemaController.singleton) {
            SchemaController.singleton = new SchemaController();
        }
        return SchemaController.singleton;
    }
    async create(req) {
        const { ws, schema } = req;
        const engine = await (0, utils_1.getWSEngine)({ ws });
        try {
            await engine.writeSchema(schema);
            return { error: null, data: undefined };
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ message: JSON.stringify(err) }),
                data: undefined,
            };
        }
    }
    async delete({ ws, id, opts, }) {
        const engine = await (0, utils_1.getWSEngine)({ ws: ws || "" });
        try {
            const data = await engine.deleteSchema(id, opts);
            return data;
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ message: JSON.stringify(err) }),
                data: undefined,
            };
        }
    }
    async query({ ws, qs }) {
        const engine = await (0, utils_1.getWSEngine)({ ws: ws || "" });
        return engine.querySchema(qs);
    }
    async update({ ws, schema, }) {
        const engine = await (0, utils_1.getWSEngine)({ ws: ws || "" });
        try {
            await engine.updateSchema(schema);
            return { error: null };
        }
        catch (err) {
            return {
                error: new common_all_1.DendronError({ message: JSON.stringify(err) }),
                data: undefined,
            };
        }
    }
}
exports.SchemaController = SchemaController;
