// @dendronhq/common-all/lib/constants/configs/global.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL = exports.ENABLE_BACK_LINKS = exports.ENABLE_CHILD_LINKS = exports.ENABLE_HASHES_FOR_FM_TAGS = exports.ENABLE_FRONTMATTER_TAGS = exports.ENABLE_NOTE_TITLE_FOR_LINK = exports.ENABLE_FM_TITLE = exports.ENABLE_KATEX = exports.ENABLE_PRETTY_REFS = exports.ENABLE_MERMAID = void 0;
const ENABLE_MERMAID = (namespace) => {
    return {
        label: `Enable Mermaid (${namespace})`,
        desc: `Enable the use of mermaid for rendering diagrams. (${namespace})`,
    };
};
exports.ENABLE_MERMAID = ENABLE_MERMAID;
const ENABLE_PRETTY_REFS = (namespace) => {
    return {
        label: `Enable Pretty Refs (${namespace})`,
        desc: `Enable rendering note references as pretty refs. (${namespace})`,
    };
};
exports.ENABLE_PRETTY_REFS = ENABLE_PRETTY_REFS;
const ENABLE_KATEX = (namespace) => {
    return {
        label: `Enable Katex (${namespace})`,
        desc: `Enable the use of katex for rendering math. (${namespace})`,
    };
};
exports.ENABLE_KATEX = ENABLE_KATEX;
const ENABLE_FM_TITLE = (namespace) => {
    return {
        label: `Enable Frontmatter Title (${namespace})`,
        desc: `Insert frontmatter title of note to the body (${namespace})`,
    };
};
exports.ENABLE_FM_TITLE = ENABLE_FM_TITLE;
const ENABLE_NOTE_TITLE_FOR_LINK = (namespace) => {
    return {
        label: `Enable Note Title for Links (${namespace})`,
        desc: `Enable rendering of naked links as the title of the note. (${namespace})`,
    };
};
exports.ENABLE_NOTE_TITLE_FOR_LINK = ENABLE_NOTE_TITLE_FOR_LINK;
const ENABLE_FRONTMATTER_TAGS = (namespace) => {
    return {
        label: `Enable Frontmatter Tags (${namespace})`,
        desc: `Show Frontmatter tags in published site. (${namespace})`,
    };
};
exports.ENABLE_FRONTMATTER_TAGS = ENABLE_FRONTMATTER_TAGS;
const ENABLE_HASHES_FOR_FM_TAGS = (namespace) => {
    return {
        label: `Enable Hashes for Frontmatter Tags (${namespace})`,
        desc: `Display a '#' symbol in front of frontmatter tags in the tags listing. (${namespace})`,
    };
};
exports.ENABLE_HASHES_FOR_FM_TAGS = ENABLE_HASHES_FOR_FM_TAGS;
const ENABLE_CHILD_LINKS = (namespace) => {
    return {
        label: `Show Child Links (${namespace})`,
        desc: `Notes will render child links (${namespace})`,
    };
};
exports.ENABLE_CHILD_LINKS = ENABLE_CHILD_LINKS;
const ENABLE_BACK_LINKS = (namespace) => {
    return {
        label: `Show Backlinks (${namespace})`,
        desc: `Notes will render backlinks (${namespace})`,
    };
};
exports.ENABLE_BACK_LINKS = ENABLE_BACK_LINKS;
exports.GLOBAL = {
    enableFMTitle: (0, exports.ENABLE_FM_TITLE)("global"),
    enableNoteTitleForLink: (0, exports.ENABLE_NOTE_TITLE_FOR_LINK)("global"),
    enableMermaid: (0, exports.ENABLE_MERMAID)("global"),
    enablePrettyRefs: (0, exports.ENABLE_PRETTY_REFS)("global"),
    enableKatex: (0, exports.ENABLE_KATEX)("global"),
    enableChildLinks: (0, exports.ENABLE_CHILD_LINKS)("global"),
    enableBackLinks: (0, exports.ENABLE_BACK_LINKS)("global"),
};
