// @dendronhq/common-all/lib/constants/configs/preview.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PREVIEW = void 0;
const global_1 = require("./global");
exports.PREVIEW = {
    enableFMTitle: (0, global_1.ENABLE_FM_TITLE)("preview"),
    enableNoteTitleForLink: (0, global_1.ENABLE_NOTE_TITLE_FOR_LINK)("preview"),
    enableFrontmatterTags: (0, global_1.ENABLE_FRONTMATTER_TAGS)("preview"),
    enableHashesForFMTags: (0, global_1.ENABLE_HASHES_FOR_FM_TAGS)("preview"),
    enableMermaid: (0, global_1.ENABLE_MERMAID)("preview"),
    enablePrettyRefs: (0, global_1.ENABLE_PRETTY_REFS)("preview"),
    enableKatex: (0, global_1.ENABLE_KATEX)("preview"),
    automaticallyShowPreview: {
        label: "Automatically Show Preview",
        desc: "Automatically show preview when opening VSCode and switching between notes.",
    },
    theme: {
        label: "The theme to use in the preview.",
        desc: "The theme to use in the preview. If unset, preview will follow your editor theme for light or dark mode. If you are using a custom theme, make sure to create the CSS file too.",
    },
};
