// @instantish/martian/build/src/notion/blocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toDo = exports.numberedListItem = exports.bulletedListItem = exports.headingThree = exports.headingTwo = exports.headingOne = exports.paragraph = void 0;
function paragraph(text) {
    return {
        object: 'block',
        type: 'paragraph',
        paragraph: {
            text: text,
        },
    };
}
exports.paragraph = paragraph;
function headingOne(text) {
    return {
        object: 'block',
        type: 'heading_1',
        heading_1: {
            text: text,
        },
    };
}
exports.headingOne = headingOne;
function headingTwo(text) {
    return {
        object: 'block',
        type: 'heading_2',
        heading_2: {
            text: text,
        },
    };
}
exports.headingTwo = headingTwo;
function headingThree(text) {
    return {
        object: 'block',
        type: 'heading_3',
        heading_3: {
            text: text,
        },
    };
}
exports.headingThree = headingThree;
function bulletedListItem(text) {
    return {
        object: 'block',
        type: 'bulleted_list_item',
        bulleted_list_item: {
            text: text,
        },
    };
}
exports.bulletedListItem = bulletedListItem;
function numberedListItem(text) {
    return {
        object: 'block',
        type: 'numbered_list_item',
        numbered_list_item: {
            text: text,
        },
    };
}
exports.numberedListItem = numberedListItem;
function toDo(checked, text) {
    return {
        object: 'block',
        type: 'to_do',
        to_do: {
            text: text,
            checked: checked,
        },
    };
}
exports.toDo = toDo;
