// @dendronhq/common-all/lib/md.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LINK_CONTENTS = exports.ALIAS_NAME = exports.LINK_NAME_NO_SPACES = exports.LINK_NAME = exports.ALIAS_DIVIDER = exports.parseAliasLink = exports.isAlias = void 0;
function isAlias(pageTitle) {
    return pageTitle.indexOf("|") !== -1;
}
exports.isAlias = isAlias;
function parseAliasLink(pageTitle) {
    const [alias, value] = pageTitle.split("|");
    return { alias, value };
}
exports.parseAliasLink = parseAliasLink;
exports.ALIAS_DIVIDER = "|";
/** A regexp fragment that matches a link name (e.g. a note name) */
exports.LINK_NAME = "[^#\\|>\\]\\[\\n]+";
exports.LINK_NAME_NO_SPACES = "[^#\\|>\\]\\[\\n\\s]+";
/** A regexp fragment that matches an alias name */
exports.ALIAS_NAME = "[^\\|>\\]\\[\\n]+"; // aliases may contain # symbols
/** A regexp fragment that matches the contents of a link (without the brackets) */
exports.LINK_CONTENTS = "" +
    // alias?
    `(` +
    `(?<alias>${exports.ALIAS_NAME}(?=\\|))` +
    "\\|" +
    ")?" +
    // name
    `(?<value>${exports.LINK_NAME})?` +
    // anchor?
    `(#(?<anchor>${exports.LINK_NAME}))?` +
    // filters?
    `(>(?<filtersRaw>.*))?`;
