// @dendronhq/dendron-cli/lib/src/commands/podsV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.enrichPodArgs = exports.setupPodArgs = void 0;
const common_all_1 = require("@dendronhq/common-all");
const pods_core_1 = require("@dendronhq/pods-core");
const lodash_1 = __importDefault(require("lodash"));
const utils_1 = require("./utils");
function setupPodArgs(args) {
    args.option("inlineConfig", {
        describe: "pass in config instead of reading from file. format is Key={key},Value={value}. If provided, this will override the value saved in the config file",
        array: true,
    });
    args.config("podConfig", "*.yml configuration file for pod", (configPath) => {
        const path = common_all_1.URI.parse(configPath);
        const configValues = pods_core_1.ConfigFileUtils.getConfigByFPath({
            fPath: path.fsPath,
        });
        if (lodash_1.default.isUndefined(configValues)) {
            throw new common_all_1.DendronError({
                message: `unable to find configuration file at ${path.fsPath}`,
            });
        }
        return { configValues };
    });
    args.option("fname", {
        describe: "full name of the note you want to export",
        type: "string",
    });
    args.option("hierarchy", {
        describe: "hierarchy you want to export",
        type: "string",
    });
    args.option("podId", {
        describe: "unique ID for your custom pod configuration.",
        type: "string",
    });
}
exports.setupPodArgs = setupPodArgs;
async function enrichPodArgs(args) {
    const { inlineConfig } = args;
    let { configValues = {} } = args;
    const engineArgs = await (0, utils_1.setupEngine)(args);
    const wsRoot = engineArgs.wsRoot;
    // return if no config is given
    if (!args.podId && !args.podConfig && !args.inlineConfig) {
        return {
            error: new common_all_1.DendronError({
                severity: common_all_1.ERROR_SEVERITY.FATAL,
                message: `no pod config found. Please provide a pod config or inline config`,
            }),
        };
    }
    // if podId is provided, get configValues from the config.{podId}.yml
    if (args.podId) {
        const podConfigPath = pods_core_1.PodUtils.getCustomConfigPath({
            wsRoot,
            podId: args.podId,
        });
        try {
            const resp = pods_core_1.ConfigFileUtils.getConfigByFPath({
                fPath: podConfigPath,
            });
            if (lodash_1.default.isUndefined(resp)) {
                return {
                    error: new common_all_1.DendronError({
                        severity: common_all_1.ERROR_SEVERITY.FATAL,
                        status: "no-custom-config",
                        message: `no pod config found for this podId. Please create a pod config at ${podConfigPath}`,
                    }),
                };
            }
            configValues = {
                ...resp,
            };
        }
        catch (err) {
            return {
                error: err,
            };
        }
    }
    // if provided, overwrite the configValues
    if (inlineConfig) {
        inlineConfig.map((conf) => {
            const [k, v] = conf.split(",");
            const key = k.split("=")[1];
            const value = v.split("=")[1];
            configValues[key] = value;
        });
    }
    // If the config has a connectionId, read the sevice connection config file.
    if (configValues.connectionId) {
        const serviceConnectionPath = pods_core_1.PodUtils.getServiceConfigPath({
            wsRoot,
            connectionId: configValues.connectionId,
        });
        try {
            const resp = pods_core_1.ConfigFileUtils.getConfigByFPath({
                fPath: serviceConnectionPath,
            });
            if (lodash_1.default.isUndefined(resp)) {
                return {
                    error: new common_all_1.DendronError({
                        status: "no-service-config",
                        message: `no service config found for this connectionId. Please create a service connection config at ${serviceConnectionPath}`,
                    }),
                };
            }
            configValues = {
                ...configValues,
                ...resp,
            };
        }
        catch (err) {
            return {
                error: err,
            };
        }
    }
    let payload;
    const { engine } = engineArgs;
    // get payload for selected export scope
    switch (configValues.exportScope) {
        case pods_core_1.PodExportScope.Workspace:
            payload = getPropsForWorkspaceScope(engine);
            break;
        case pods_core_1.PodExportScope.Vault:
            payload = getPropsForVaultScope({ engine, vaultName: args.vault });
            break;
        case pods_core_1.PodExportScope.Note:
            payload = getPropsForNoteScope({
                engine,
                vaultName: args.vault,
                fname: args.fname,
            });
            break;
        case pods_core_1.PodExportScope.Hierarchy:
            payload = getPropsForHierarchyScope({
                engine,
                hierarchy: args.hierarchy,
                vaultName: args.vault,
            });
            break;
        default:
            throw new common_all_1.DendronError({
                message: `the CLI doesn't support the provided export scope: ${configValues.exportScope}. please run this export pod using the Dendon plugin`,
            });
    }
    return {
        data: {
            ...args,
            ...engineArgs,
            config: configValues,
            payload,
        },
    };
}
exports.enrichPodArgs = enrichPodArgs;
/**
 *
 * @param engine
 * @returns all notes in workspace
 */
const getPropsForWorkspaceScope = (engine) => {
    return Object.values(engine.notes).filter((notes) => notes.stub !== true);
};
/**
 *
 * @returns all notes in the vault
 */
const getPropsForVaultScope = (opts) => {
    const { engine, vaultName } = opts;
    const vault = checkVaultArgs({ engine, vaultName });
    return Object.values(engine.notes).filter((note) => note.stub !== true && common_all_1.VaultUtils.isEqualV2(note.vault, vault));
};
const getPropsForNoteScope = (opts) => {
    const { engine, fname, vaultName } = opts;
    const vault = checkVaultArgs({ engine, vaultName });
    if (!fname) {
        throw new common_all_1.DendronError({
            message: "Please provide fname of note in --fname arg",
        });
    }
    const note = common_all_1.NoteUtils.getNoteByFnameFromEngine({ fname, vault, engine });
    if (!note)
        throw new common_all_1.DendronError({
            message: `Cannot find note with fname ${fname} in vault ${vault}`,
        });
    return [note];
};
// returns notes within a hierarchy (for a specefic vault)
const getPropsForHierarchyScope = (opts) => {
    const { engine, hierarchy, vaultName } = opts;
    if (!hierarchy) {
        throw new common_all_1.DendronError({
            message: "Please provide hierarchy in --hierarchy arg",
        });
    }
    const vault = checkVaultArgs({ engine, vaultName });
    return Object.values(engine.notes).filter((value) => value.fname.startsWith(hierarchy) &&
        value.stub !== true &&
        common_all_1.VaultUtils.isEqualV2(value.vault, vault));
};
/**
 * This method check --vault argument. For a single vault workspace, if --vault not provided,
 * returns the vault from workspace.
 * For multi-vault workspace, if no --vault is given, returns an error, else returns selected vault
 */
const checkVaultArgs = (opts) => {
    const { engine, vaultName } = opts;
    const { vaults } = engine;
    if (lodash_1.default.size(vaults) > 1 && !vaultName) {
        throw new common_all_1.DendronError({
            message: "Please provide vault name in --vault arg",
        });
    }
    else {
        return vaultName
            ? common_all_1.VaultUtils.getVaultByNameOrThrow({ vaults, vname: vaultName })
            : vaults[0];
    }
};
