// @dendronhq/common-all/lib/types/configs/publishing/publishing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genDefaultPublishingConfig = exports.DuplicateNoteActionEnum = exports.Theme = void 0;
const github_1 = require("./github");
const seo_1 = require("./seo");
var Theme;
(function (Theme) {
    Theme["DARK"] = "dark";
    Theme["LIGHT"] = "light";
    Theme["CUSTOM"] = "custom";
})(Theme = exports.Theme || (exports.Theme = {}));
var DuplicateNoteActionEnum;
(function (DuplicateNoteActionEnum) {
    DuplicateNoteActionEnum["useVault"] = "useVault";
})(DuplicateNoteActionEnum = exports.DuplicateNoteActionEnum || (exports.DuplicateNoteActionEnum = {}));
/**
 * Generate default {@link DendronPublishingConfig}
 * @returns DendronPublishingConfig
 */
function genDefaultPublishingConfig() {
    return {
        enableFMTitle: true,
        enableNoteTitleForLink: true,
        enableMermaid: true,
        enablePrettyRefs: true,
        enableKatex: true,
        copyAssets: true,
        siteHierarchies: ["root"],
        writeStubs: false,
        siteRootDir: "docs",
        seo: (0, seo_1.genDefaultSEOConfig)(),
        github: (0, github_1.genDefaultGithubConfig)(),
        enableSiteLastModified: true,
        enableFrontmatterTags: true,
        enableHashesForFMTags: false,
        enableRandomlyColoredTags: true,
        enableTaskNotes: true,
        enablePrettyLinks: true,
    };
}
exports.genDefaultPublishingConfig = genDefaultPublishingConfig;
