// @dendronhq/pods-core/lib/v2/pods/export/AirtableExportPodV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirtableExportPodV2 = void 0;
const common_all_1 = require("@dendronhq/common-all");
const common_server_1 = require("@dendronhq/common-server");
const limiter_1 = require("limiter");
const lodash_1 = __importDefault(require("lodash"));
const __1 = require("../../..");
class AirtableUtilsV2 {
    /***
     * Chunk all calls into records of 10 (Airtable API limit and call using limiter)
     */
    static async chunkAndCall(allRecords, limiter, func) {
        const chunks = lodash_1.default.chunk(allRecords, 10);
        const errors = [];
        const out = await Promise.all(chunks.flatMap(async (record) => {
            await limiter.removeTokens(1);
            try {
                const _records = await func(record);
                return _records;
            }
            catch (error) {
                let message;
                if ((error === null || error === void 0 ? void 0 : error.statusCode) === common_all_1.StatusCodes.UNPROCESSABLE_ENTITY) {
                    const airtableError = error;
                    if (airtableError.error === "INVALID_MULTIPLE_CHOICE_OPTIONS") {
                        // example airtable error message: 'Insufficient permissions to create new select option ""scope.xyz""'
                        const value = airtableError.message.split('""')[1];
                        const field = lodash_1.default.findKey(record[0].fields, lodash_1.default.partial(lodash_1.default.isEqual, value));
                        message = field
                            ? `The choice ${value} for field ${field} does not exactly match with an existing option. Please check what values are allowed in Airtable`
                            : airtableError.message;
                    }
                    else if (airtableError.error === "INVALID_VALUE_FOR_COLUMN") {
                        message = airtableError.message;
                    }
                }
                const _error = new common_all_1.DendronError({
                    innerError: error,
                    payload: record,
                    message: `Error during Airtable Export. ${message}`,
                });
                errors.push(_error);
                return;
            }
        }));
        return {
            data: lodash_1.default.flatten(out).filter((ent) => !lodash_1.default.isUndefined(ent)),
            errors,
        };
    }
}
/**
 * Airtable Export Pod (V2 - for compatibility with Pod V2 workflow). This pod
 * will export data to a table row in Airtable.
 */
class AirtableExportPodV2 {
    constructor({ airtable, config, engine }) {
        this._airtableBase = airtable.base(config.baseId);
        this._config = config;
        this._engine = engine;
    }
    cleanNotes(notes, fnameFilters) {
        return lodash_1.default.reject(notes, (ent) => {
            return lodash_1.default.some(fnameFilters, (pat) => (0, common_all_1.minimatch)(ent.fname, pat));
        });
    }
    async exportNotes(input) {
        input = this.cleanNotes(input, lodash_1.default.get(this._config, "filters.fname"));
        const resp = this.getPayloadForNotes(input);
        if (resp.error) {
            return {
                data: {},
                error: resp.error,
            };
        }
        const { create, update } = resp.data;
        const limiter = new limiter_1.RateLimiter({
            tokensPerInterval: 5,
            interval: "second",
        });
        const createRequest = await AirtableUtilsV2.chunkAndCall(create, limiter, this._airtableBase(this._config.tableName).create);
        const updateRequest = await AirtableUtilsV2.chunkAndCall(update, limiter, this._airtableBase(this._config.tableName).update);
        const errors = createRequest.errors.concat(updateRequest.errors);
        const data = {
            created: createRequest.data,
            updated: updateRequest.data,
        };
        if (errors.length > 0) {
            return {
                data,
                error: new common_all_1.DendronCompositeError(errors),
            };
        }
        else {
            return common_all_1.ResponseUtil.createHappyResponse({
                data,
            });
        }
    }
    /**
     * Get mapping of fields that will be updated in airtable
     * @param notes
     * @returns
     */
    getPayloadForNotes(notes) {
        const logger = (0, common_server_1.createLogger)("AirtablePublishPodV2");
        const resp = __1.AirtableUtils.notesToSrcFieldMap({
            notes,
            srcFieldMapping: this._config.sourceFieldMapping,
            logger,
            engine: this._engine,
            podId: this._config.podId,
        });
        return resp;
    }
    static config() {
        return __1.ConfigFileUtils.createExportConfig({
            required: ["connectionId", "baseId", "tableName", "sourceFieldMapping"],
            properties: {
                connectionId: {
                    description: "ID of the Airtable Connected Service",
                    type: "string",
                },
                baseId: {
                    description: "airtable base id",
                    type: "string",
                },
                tableName: { type: "string", description: "Name of the airtable" },
                sourceFieldMapping: {
                    type: "object",
                    required: [],
                    description: "mapping of airtable fields with the note eg: {Created On: created, Notes: body}",
                },
                filters: {
                    type: "object",
                    required: [],
                    nullable: true,
                    properties: {
                        fname: {
                            type: "array",
                            items: {
                                type: "string",
                            },
                        },
                    },
                },
            },
        });
    }
}
exports.AirtableExportPodV2 = AirtableExportPodV2;
