// @dendronhq/pods-core/lib/v2/pods/export/GoogleDocsExportPodV2.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleDocsUtils = exports.GoogleDocsExportPodV2 = void 0;
const common_all_1 = require("@dendronhq/common-all");
const form_data_1 = __importDefault(require("form-data"));
const limiter_1 = require("limiter");
const lodash_1 = __importDefault(require("lodash"));
const __1 = require("../../..");
/**
 * GDoc Export Pod (V2 - for compatibility with Pod V2 workflow). Supports only
 * exportNote() for now
 */
class GoogleDocsExportPodV2 {
    constructor({ podConfig, engine, port, }) {
        this._config = podConfig;
        this._engine = engine;
        this._vaults = engine.vaults;
        this._wsRoot = engine.wsRoot;
        this._port = port;
    }
    async exportNotes(notes) {
        const resp = await this.getPayloadForNotes(notes);
        let { accessToken } = this._config;
        const { expirationTime, refreshToken, parentFolderId } = this._config;
        try {
            accessToken = await this.checkTokenExpiry(expirationTime, accessToken, refreshToken);
        }
        catch (err) {
            return {
                data: {},
                error: err,
            };
        }
        /**
         * The rate of Drive API write requests is limited—avoid exceeding 3 requests per second
         * of sustained write or insert requests
         */
        const limiter = new limiter_1.RateLimiter({
            tokensPerInterval: 3,
            interval: "second",
        });
        const docToCreate = resp.filter((note) => lodash_1.default.isUndefined(note.documentId));
        const docToUpdate = resp.filter((note) => !lodash_1.default.isUndefined(note.documentId));
        const createRequest = await this.createGdoc({
            docToCreate,
            accessToken,
            limiter,
            parentFolderId,
        });
        const updateRequest = await this.overwriteGdoc({
            docToUpdate,
            accessToken,
            limiter,
        });
        const errors = createRequest.errors.concat(updateRequest.errors);
        const data = {
            created: createRequest.data,
            updated: updateRequest.data,
        };
        if (errors.length > 0) {
            return {
                data,
                error: new common_all_1.DendronCompositeError(errors),
            };
        }
        else {
            return common_all_1.ResponseUtil.createHappyResponse({
                data,
            });
        }
    }
    /**
     * Method to check if the accessToken is valid, if not returns a refreshed accessToken
     */
    async checkTokenExpiry(expirationTime, accessToken, refreshToken) {
        if (common_all_1.Time.now().toSeconds() > expirationTime) {
            accessToken = await __1.PodUtils.refreshGoogleAccessToken(refreshToken, this._port, this._config.connectionId);
        }
        return accessToken;
    }
    /**
     * Method to return the payload for creating/overwriting a google document.
     * @param notes
     * @returns an array of payload for each note.
     */
    getPayloadForNotes(notes) {
        return Promise.all(notes.map(async (input) => {
            const pod = new __1.HTMLPublishPod();
            const config = {
                fname: input.fname,
                vaultName: input.vault,
                dest: "stdout",
                convertLinks: false,
                enablePrettyRefs: false,
            };
            // converts markdown to html using HTMLPublish pod. The Drive API supports converting MIME types while creating a file.
            let data = await pod.plant({
                config,
                engine: this._engine,
                note: input,
                vaults: this._vaults,
                wsRoot: this._wsRoot,
            });
            const blockquote = `<blockquote>This note was exported from [[${input.fname}]] with <a href="https://dendron.so/">Dendron</a></blockquote>`;
            // wrap data in html tags
            data = `<html>${blockquote}${data}</html>`;
            const content = Buffer.from(data);
            const documentId = input.custom.documentId;
            return {
                content,
                documentId,
                name: input.fname,
                dendronId: input.id,
            };
        }));
    }
    /**
     * Creates new google documents for given notes.
     */
    async createGdoc(opts) {
        const { docToCreate, accessToken, limiter, parentFolderId = "root" } = opts;
        const errors = [];
        const out = await Promise.all(docToCreate.map(async ({ name, content, dendronId }) => {
            await (limiter === null || limiter === void 0 ? void 0 : limiter.removeTokens(1));
            try {
                let revisionId = "";
                //metadata is used by drive API to understand the required MIME type
                const metadata = {
                    name,
                    mimeType: "application/vnd.google-apps.document",
                    parents: [`${parentFolderId}`],
                };
                const formData = new form_data_1.default();
                formData.append("metadata", JSON.stringify(metadata), {
                    contentType: "application/json",
                });
                formData.append("file", content);
                const response = await (0, common_all_1.axios)({
                    method: "POST",
                    url: "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&supportsAllDrives=true",
                    headers: {
                        Authorization: `Bearer ${accessToken}`,
                        "Content-Type": `multipart/related; boundary=${formData.getBoundary()}`,
                    },
                    data: formData,
                });
                if (response) {
                    revisionId = await this.getRevisionId({
                        documentId: response.data.id,
                        accessToken,
                    });
                }
                return {
                    documentId: response.data.id,
                    revisionId,
                    dendronId,
                };
            }
            catch (err) {
                errors.push(err);
                return;
            }
        }));
        return {
            data: out,
            errors,
        };
    }
    /**
     * If a note has document id, overwrite the existing gdoc with the note's content.
     * @param opts
     * @returns
     */
    async overwriteGdoc(opts) {
        const { docToUpdate, accessToken, limiter } = opts;
        const errors = [];
        const out = await Promise.all(docToUpdate.map(async ({ content, documentId, dendronId }) => {
            if (!documentId)
                return;
            await limiter.removeTokens(1);
            try {
                const fileSize = content.length;
                let revisionId = "";
                const response = await (0, common_all_1.axios)({
                    method: "PUT",
                    url: `https://www.googleapis.com/upload/drive/v2/files/${documentId}`,
                    headers: {
                        Authorization: `Bearer ${accessToken}`,
                        "Content-Type": "application/json",
                        "Content-Range": `bytes 0-${fileSize - 1}/${fileSize}`,
                    },
                    data: content,
                });
                if (response) {
                    revisionId = await this.getRevisionId({ documentId, accessToken });
                }
                return {
                    documentId: response.data.id,
                    revisionId,
                    dendronId,
                };
            }
            catch (err) {
                errors.push(err);
                return;
            }
        }));
        return {
            data: out,
            errors,
        };
    }
    /**
     * Method to retrieve revisionId of a document. The drive api only returns document id in response.
     */
    async getRevisionId(opts) {
        const { accessToken, documentId } = opts;
        try {
            const result = await common_all_1.axios.get(`https://docs.googleapis.com/v1/documents/${documentId}`, {
                headers: {
                    Authorization: `Bearer ${accessToken}`,
                    "Content-Type": "application/json",
                },
            });
            return result.data.revisionId;
        }
        catch (err) {
            throw new common_all_1.DendronError({ message: (0, common_all_1.stringifyError)(err) });
        }
    }
    static config() {
        return __1.ConfigFileUtils.createExportConfig({
            required: ["connectionId"],
            properties: {
                connectionId: {
                    description: "ID of the Google Connected Service",
                    type: "string",
                },
                parentFolderId: {
                    description: "ID of parent folder in google drive",
                    type: "string",
                    nullable: true,
                },
            },
        });
    }
}
exports.GoogleDocsExportPodV2 = GoogleDocsExportPodV2;
class GoogleDocsUtils {
    static async updateNotesWithCustomFrontmatter(records, engine, parentFolderId) {
        await Promise.all(records.map(async (record) => {
            if (lodash_1.default.isUndefined(record))
                return;
            const { documentId, revisionId, dendronId } = record;
            if (!dendronId)
                return;
            const note = engine.notes[dendronId];
            note.custom = {
                ...note.custom,
                documentId,
                revisionId,
                uri: `https://docs.google.com/document/d/${documentId}/edit`,
                parentFolderId,
            };
            await engine.writeNote(note);
        }));
    }
}
exports.GoogleDocsUtils = GoogleDocsUtils;
