// @dendronhq/common-all/lib/store/SchemaMetadataStore.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaMetadataStore = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const error_1 = require("../error");
class SchemaMetadataStore {
    constructor() {
        this._schemaMetadataById = {};
    }
    /**
     * See {@link IDataStore.get}
     */
    async get(key) {
        const maybeSchema = this._schemaMetadataById[key];
        if (maybeSchema) {
            return { data: lodash_1.default.cloneDeep(maybeSchema) };
        }
        else {
            return {
                error: error_1.DendronError.createFromStatus({
                    status: constants_1.ERROR_STATUS.CONTENT_NOT_FOUND,
                    message: `SchemaModuleProps not found for key ${key}.`,
                    severity: constants_1.ERROR_SEVERITY.MINOR,
                }),
            };
        }
    }
    async find(_opts) {
        throw new Error("Method not implemented.");
    }
    /**
     * See {@link IDataStore.write}
     *
     * Add schema to _schemaMetadataById. If schema root id already exists, override existing schema
     */
    async write(key, data) {
        this._schemaMetadataById[data.root.id] = data;
        return { data: key };
    }
    /**
     * See {@link IDataStore.delete}
     *
     * Remove schema from both _schemaMetadataById.
     */
    async delete(key) {
        delete this._schemaMetadataById[key];
        return { data: key };
    }
}
exports.SchemaMetadataStore = SchemaMetadataStore;
