// @dendronhq/common-all/lib/types/lookup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LookupFilterTypeEnum = exports.LookupSplitTypeEnum = exports.LookupEffectTypeEnum = exports.LookupNoteTypeEnum = exports.LookupSelectionTypeEnum = void 0;
var LookupSelectionTypeEnum;
(function (LookupSelectionTypeEnum) {
    LookupSelectionTypeEnum["selection2link"] = "selection2link";
    LookupSelectionTypeEnum["selectionExtract"] = "selectionExtract";
    LookupSelectionTypeEnum["selection2Items"] = "selection2Items";
    LookupSelectionTypeEnum["none"] = "none";
})(LookupSelectionTypeEnum = exports.LookupSelectionTypeEnum || (exports.LookupSelectionTypeEnum = {}));
var LookupNoteTypeEnum;
(function (LookupNoteTypeEnum) {
    LookupNoteTypeEnum["journal"] = "journal";
    LookupNoteTypeEnum["scratch"] = "scratch";
    LookupNoteTypeEnum["task"] = "task";
    LookupNoteTypeEnum["none"] = "none";
})(LookupNoteTypeEnum = exports.LookupNoteTypeEnum || (exports.LookupNoteTypeEnum = {}));
var LookupEffectTypeEnum;
(function (LookupEffectTypeEnum) {
    LookupEffectTypeEnum["copyNoteLink"] = "copyNoteLink";
    LookupEffectTypeEnum["multiSelect"] = "multiSelect";
})(LookupEffectTypeEnum = exports.LookupEffectTypeEnum || (exports.LookupEffectTypeEnum = {}));
var LookupSplitTypeEnum;
(function (LookupSplitTypeEnum) {
    LookupSplitTypeEnum["horizontal"] = "horizontal";
})(LookupSplitTypeEnum = exports.LookupSplitTypeEnum || (exports.LookupSplitTypeEnum = {}));
var LookupFilterTypeEnum;
(function (LookupFilterTypeEnum) {
    LookupFilterTypeEnum["directChildOnly"] = "directChildOnly";
})(LookupFilterTypeEnum = exports.LookupFilterTypeEnum || (exports.LookupFilterTypeEnum = {}));
