// @dendronhq/common-all/lib/types/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyDuplicateNoteAction = exports.LegacyInsertNoteLinkAliasMode = exports.LegacyLookupSelectionType = exports.LegacyNoteAddBehavior = exports.InstallStatus = exports.WorkspaceType = exports.DVaultSync = exports.DVaultVisibility = void 0;
var DVaultVisibility;
(function (DVaultVisibility) {
    DVaultVisibility["PRIVATE"] = "private";
})(DVaultVisibility = exports.DVaultVisibility || (exports.DVaultVisibility = {}));
var DVaultSync;
(function (DVaultSync) {
    DVaultSync["SKIP"] = "skip";
    DVaultSync["NO_PUSH"] = "noPush";
    DVaultSync["NO_COMMIT"] = "noCommit";
    DVaultSync["SYNC"] = "sync";
})(DVaultSync = exports.DVaultSync || (exports.DVaultSync = {}));
var WorkspaceType;
(function (WorkspaceType) {
    WorkspaceType["NATIVE"] = "NATIVE";
    WorkspaceType["CODE"] = "CODE";
    WorkspaceType["NONE"] = "NONE";
})(WorkspaceType = exports.WorkspaceType || (exports.WorkspaceType = {}));
/**
 * Extension Install Status
 */
var InstallStatus;
(function (InstallStatus) {
    InstallStatus["NO_CHANGE"] = "NO_CHANGE";
    InstallStatus["INITIAL_INSTALL"] = "INITIAL_INSTALL";
    InstallStatus["UPGRADED"] = "UPGRADED";
})(InstallStatus = exports.InstallStatus || (exports.InstallStatus = {}));
var LegacyNoteAddBehavior;
(function (LegacyNoteAddBehavior) {
    LegacyNoteAddBehavior["childOfDomain"] = "childOfDomain";
    LegacyNoteAddBehavior["childOfDomainNamespace"] = "childOfDomainNamespace";
    LegacyNoteAddBehavior["childOfCurrent"] = "childOfCurrent";
    LegacyNoteAddBehavior["asOwnDomain"] = "asOwnDomain";
})(LegacyNoteAddBehavior = exports.LegacyNoteAddBehavior || (exports.LegacyNoteAddBehavior = {}));
var LegacyLookupSelectionType;
(function (LegacyLookupSelectionType) {
    LegacyLookupSelectionType["selection2link"] = "selection2link";
    LegacyLookupSelectionType["selectionExtract"] = "selectionExtract";
    LegacyLookupSelectionType["none"] = "none";
})(LegacyLookupSelectionType = exports.LegacyLookupSelectionType || (exports.LegacyLookupSelectionType = {}));
var LegacyInsertNoteLinkAliasMode;
(function (LegacyInsertNoteLinkAliasMode) {
    LegacyInsertNoteLinkAliasMode["snippet"] = "snippet";
    LegacyInsertNoteLinkAliasMode["selection"] = "selection";
    LegacyInsertNoteLinkAliasMode["title"] = "title";
    LegacyInsertNoteLinkAliasMode["prompt"] = "prompt";
    LegacyInsertNoteLinkAliasMode["none"] = "none";
})(LegacyInsertNoteLinkAliasMode = exports.LegacyInsertNoteLinkAliasMode || (exports.LegacyInsertNoteLinkAliasMode = {}));
var LegacyDuplicateNoteAction;
(function (LegacyDuplicateNoteAction) {
    LegacyDuplicateNoteAction["USE_VAULT"] = "useVault";
})(LegacyDuplicateNoteAction = exports.LegacyDuplicateNoteAction || (exports.LegacyDuplicateNoteAction = {}));
