// @dendronhq/common-server/lib/errorReporting.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventModifier = exports.initializeSentry = exports.isBadErrorThatShouldBeSampled = exports.rewriteFilename = void 0;
const common_all_1 = require("@dendronhq/common-all");
const integrations_1 = require("@sentry/integrations");
const Sentry = __importStar(require("@sentry/node"));
// Extracted to make testing easy
function rewriteFilename(filename) {
    // Convert backslash to forward slash; Sentry should be able to handle the rest of the formatting:
    filename = filename.split("\\").join("/");
    // Replace windows C: with nothing
    filename = filename.replace(/^[A-Za-z]:/, "");
    // Remove everything including the dendron directory, this is usually stuff like '/Users/someone/...'
    // We have to do two regexes because doing dendron(\.[A-Za-z]*-[0-9.]*)? does not work properly
    const prodRegex = /^\/.*dendron\.[A-Za-z_]*-[0-9.]*\//;
    const devRegex = /^\/.*dendron\//;
    const prefix = "app:///";
    const newFilename = filename.replace(prodRegex, prefix);
    if (newFilename !== filename) {
        return newFilename;
    }
    else {
        return filename.replace(devRegex, prefix);
    }
}
exports.rewriteFilename = rewriteFilename;
// This offers some protection against accidentally sending too many errors to Sentry
const BAD_ERROR_SAMPLE_RATE = 0.001;
function isBadErrorThatShouldBeSampled(error) {
    return (error &&
        typeof error !== "string" &&
        error.message &&
        error.message.includes("ENOENT: no such file or directory"));
}
exports.isBadErrorThatShouldBeSampled = isBadErrorThatShouldBeSampled;
/**
 * Initialize Sentry
 * @param environment
 * @returns
 *  ^4wcl13fw6gub
 */
function initializeSentry({ environment, sessionId, release, }) {
    const dsn = "https://bc206b31a30a4595a2efb31e8cc0c04e@o949501.ingest.sentry.io/5898219";
    const initialScope = {};
    if (sessionId) {
        initialScope.tags = { sessionId };
    }
    Sentry.init({
        dsn,
        defaultIntegrations: false,
        // Error stack trace sample rate: send all errors to sentry
        sampleRate: 1.0,
        // Transaction sample rate. Transactions are activities like page loads and api calls
        // The configuration property name is a bit misleading. We don't use them right now.
        tracesSampleRate: 0.0,
        enabled: true,
        environment,
        release,
        attachStacktrace: true,
        beforeSend: eventModifier,
        initialScope,
        integrations: [
            new integrations_1.RewriteFrames({
                iteratee: (frame) => {
                    if (!frame.filename) {
                        return frame;
                    }
                    frame.filename = rewriteFilename(frame.filename);
                    return frame;
                },
            }),
        ],
    });
    return;
}
exports.initializeSentry = initializeSentry;
function eventModifier(event, hint) {
    var _a;
    const error = hint === null || hint === void 0 ? void 0 : hint.originalException;
    if (isBadErrorThatShouldBeSampled(error) &&
        Math.random() > BAD_ERROR_SAMPLE_RATE) {
        return null;
    }
    // Add more information to the event extras payload:
    if (error && error instanceof common_all_1.DendronError) {
        // This is a bit hacky because it overwrites the existing extra context
        // TODO: figure out how to handle contexts in a uniform way
        event.extra = {
            ...event.extra,
            name: error.name,
            message: error.message,
            payload: error.payload,
            severity: (_a = error.severity) === null || _a === void 0 ? void 0 : _a.toString(),
            code: error.code,
            status: error.status,
            innerError: error,
        };
    }
    return event;
}
exports.eventModifier = eventModifier;
