// @dendronhq/common-server/lib/filesv2.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileUtils = exports.tmp = exports.FileExtensionUtils = exports.moveIfExists = exports.isSelfContainedVaultFolder = exports.findNonNoteFile = exports.fileExists = exports.dot2Slash = exports.writeJSONWithComments = exports.writeJSONWithCommentsSync = exports.pathForVaultRoot = exports.vault2Path = exports.tmpDir = exports.readJSONWithCommentsSync = exports.readJSONWithComments = exports.assignJSONWithComment = exports.schemaModuleProps2File = exports.schemaModuleOpts2File = exports.note2File = exports.uniqueOutermostFolders = exports.isInsidePath = exports.findDownTo = exports.WS_FILE_MAX_SEARCH_DEPTH = exports.findUpTo = exports.goUpTo = exports.note2String = exports.file2Note = exports.string2Schema = exports.file2Schema = exports.createFileWatcher = exports.COMMON_FOLDER_IGNORES = void 0;
const common_all_1 = require("@dendronhq/common-all");
const anymatch_1 = __importDefault(require("anymatch"));
const comment_json_1 = require("comment-json");
const fs_extra_1 = __importDefault(require("fs-extra"));
const js_yaml_1 = __importStar(require("js-yaml"));
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
// @ts-ignore
const tmp_1 = __importStar(require("tmp"));
exports.tmp = tmp_1.default;
const files_1 = require("./files");
const parser_1 = require("./parser");
const textextensions_1 = __importDefault(require("textextensions"));
/** Dendron should ignore any of these folders when watching or searching folders.
 *
 * These folders are unlikely to contain anything Dendron would like to find, so we can ignore them.
 *
 * Example usage:
 * ```ts
 * if (!anymatch(COMMON_FOLDER_IGNORES, folder)) {
 *   // Good folder!
 * }
 * ```
 */
exports.COMMON_FOLDER_IGNORES = [
    "**/.*/**",
    "**/node_modules/**",
    "**/.git/**",
    "**/__pycache__/**", // python
];
async function createFileWatcher(opts) {
    const { numTries, fpath, onChange } = lodash_1.default.defaults(opts, {
        numTries: 5,
    });
    const didCreate = false;
    return new Promise((resolve, _reject) => {
        if (!fs_extra_1.default.existsSync(fpath)) {
            return setTimeout(() => {
                resolve(_createFileWatcher({
                    ...opts,
                    numTries: numTries - 1,
                    isCreate: true,
                }));
            }, 3000);
        }
        const watcher = fs_extra_1.default.watch(fpath, () => {
            onChange({ fpath });
        });
        return resolve({ watcher, didCreate });
    });
}
exports.createFileWatcher = createFileWatcher;
async function _createFileWatcher(opts) {
    const { numTries, fpath, onChange, onCreate } = lodash_1.default.defaults(opts, {
        numTries: 5,
    });
    if (numTries <= 0) {
        throw new common_all_1.DendronError({ message: "exceeded numTries" });
    }
    return new Promise(async (resolve, _reject) => {
        if (!fs_extra_1.default.existsSync(fpath)) {
            console.log({ fpath, msg: "not exist" });
            return setTimeout(() => {
                resolve(createFileWatcher({ ...opts, numTries: numTries - 1 }));
            }, 3000);
        }
        await onCreate({ fpath });
        const watcher = fs_extra_1.default.watch(fpath, () => {
            onChange({ fpath });
        });
        return resolve({ watcher, didCreate: true });
    });
}
async function file2Schema(fpath, wsRoot) {
    const root = { fsPath: path_1.default.dirname(fpath) };
    const fname = path_1.default.basename(fpath, ".schema.yml");
    const schemaOpts = js_yaml_1.default.load(await fs_extra_1.default.readFile(fpath, { encoding: "utf8" }));
    return parser_1.SchemaParserV2.parseRaw(schemaOpts, { root, fname, wsRoot });
}
exports.file2Schema = file2Schema;
async function string2Schema({ vault, content, fname, wsRoot, }) {
    const schemaOpts = js_yaml_1.default.load(content);
    return parser_1.SchemaParserV2.parseRaw(schemaOpts, {
        root: vault,
        fname,
        wsRoot,
    });
}
exports.string2Schema = string2Schema;
// TODO: consider throwing error if no frontmatter
function file2Note(fpath, vault, toLowercase) {
    if (!fs_extra_1.default.existsSync(fpath)) {
        const error = common_all_1.DendronError.createFromStatus({
            status: common_all_1.ERROR_STATUS.INVALID_STATE,
            message: `${fpath} does not exist`,
        });
        return {
            error,
        };
    }
    else {
        const content = fs_extra_1.default.readFileSync(fpath, { encoding: "utf8" });
        const { name } = path_1.default.parse(fpath);
        const fname = toLowercase ? name.toLowerCase() : name;
        return {
            data: (0, common_all_1.string2Note)({ content, fname, vault }),
        };
    }
}
exports.file2Note = file2Note;
/** Read the contents of a note from the filesystem.
 *
 * Warning! The note contents may be out of date compared to changes in the editor.
 * Consider using `NoteUtils.serialize` instead.
 */
function note2String(opts) {
    const notePath = common_all_1.NoteUtils.getFullPath(opts);
    return fs_extra_1.default.readFile(notePath, { encoding: "utf8" });
}
exports.note2String = note2String;
/**
 * Go to dirname that {fname} is contained in
 * @param maxLvl? - default: 10
 @deprecated use {@link findUpTo}
 */
function goUpTo(opts) {
    var _b;
    const { fname, base } = opts;
    let maxLvl = (_b = opts.maxLvl) !== null && _b !== void 0 ? _b : 10;
    const lvls = [];
    while (maxLvl > 0) {
        const tryPath = path_1.default.join(base, ...lvls, fname);
        if (fs_extra_1.default.existsSync(tryPath)) {
            return path_1.default.dirname(tryPath);
        }
        maxLvl -= 1;
        lvls.push("..");
    }
    throw Error(`no root found from ${base}`);
}
exports.goUpTo = goUpTo;
/**
 * Go to dirname that {fname} is contained in, going out (up the tree) from base.
 * @param maxLvl - default: 3
 * @param returnDirPath - return path to directory, default: false
 */
function findUpTo(opts) {
    const { fname, base, maxLvl, returnDirPath } = lodash_1.default.defaults(opts, {
        maxLvl: 3,
        returnDirPath: false,
    });
    const lvls = [];
    let acc = 0;
    while (maxLvl - acc > 0) {
        const tryPath = path_1.default.join(base, ...lvls, fname);
        if (fs_extra_1.default.existsSync(tryPath)) {
            return returnDirPath ? path_1.default.dirname(tryPath) : tryPath;
        }
        acc += 1;
        lvls.push("..");
    }
    return undefined;
}
exports.findUpTo = findUpTo;
exports.WS_FILE_MAX_SEARCH_DEPTH = 3;
/**
 * Go to dirname that {fname} is contained in, going in (deeper into tree) from base.
 * @param maxLvl Default 3, how deep to go down in the file tree. Keep in mind that the tree gets wider and this search becomes exponentially more expensive the deeper we go.
 * @param returnDirPath - return path to directory, default: false
 *
 * One warning: this will not search into folders starting with `.` to avoid searching through things like the `.git` folder.
 */
async function findDownTo(opts) {
    const { fname, base, maxLvl, returnDirPath } = {
        maxLvl: exports.WS_FILE_MAX_SEARCH_DEPTH,
        returnDirPath: false,
        ...opts,
    };
    const contents = await fs_extra_1.default.readdir(base);
    let found = contents.filter((foundFile) => foundFile === fname)[0];
    if (found) {
        found = path_1.default.join(base, found);
        return returnDirPath ? path_1.default.dirname(found) : found;
    }
    if (maxLvl > 1) {
        // Keep searching recursively
        return (await Promise.all(contents.map(async (folder) => {
            // Find the folders in the current folder
            const subfolder = await fs_extra_1.default.stat(path_1.default.join(base, folder));
            if (!subfolder.isDirectory())
                return;
            // Exclude folders starting with . to skip stuff like `.git`
            if ((0, anymatch_1.default)(exports.COMMON_FOLDER_IGNORES, folder))
                return;
            return findDownTo({
                ...opts,
                base: path_1.default.join(base, folder),
                maxLvl: maxLvl - 1,
            });
        }))).filter(common_all_1.isNotUndefined)[0];
    }
    return undefined;
}
exports.findDownTo = findDownTo;
/** Returns true if `inner` is inside of `outer`, and false otherwise.
 *
 * If `inner === outer`, then that also returns false.
 */
function isInsidePath(outer, inner) {
    // When going from `outer` to `inner`
    const relPath = path_1.default.relative(outer, inner);
    // If we have to leave `outer`, or if we have to switch to a
    // different drive with an absolute path, then `inner` can't be
    // inside `outer` (or `inner` and `outer` are identical)
    return (!relPath.startsWith("..") && !path_1.default.isAbsolute(relPath) && relPath !== "");
}
exports.isInsidePath = isInsidePath;
/** Returns the list of unique, outermost folders. No two folders returned are nested within each other. */
function uniqueOutermostFolders(folders) {
    // Avoid duplicates
    folders = lodash_1.default.uniq(folders);
    if (folders.length === 1)
        return folders;
    return folders.filter((currentFolder) => folders.every((otherFolder) => {
        // `currentFolder` is not inside any other folder
        return !isInsidePath(otherFolder, currentFolder);
    }));
}
exports.uniqueOutermostFolders = uniqueOutermostFolders;
/**
 * Return hash of written file
 */
async function note2File({ note, vault, wsRoot, }) {
    const { fname } = note;
    const ext = ".md";
    const payload = common_all_1.NoteUtils.serialize(note, { excludeStub: true });
    const vpath = (0, exports.vault2Path)({ vault, wsRoot });
    await fs_extra_1.default.writeFile(path_1.default.join(vpath, fname + ext), payload);
    return (0, common_all_1.genHash)(payload);
}
exports.note2File = note2File;
function serializeModuleOpts(moduleOpts) {
    const { version, imports, schemas } = lodash_1.default.defaults(moduleOpts, {
        imports: [],
    });
    const out = {
        version,
        imports,
        schemas: lodash_1.default.values(schemas).map((ent) => common_all_1.SchemaUtils.serializeSchemaProps(ent)),
    };
    return js_yaml_1.default.dump(out, { schema: js_yaml_1.JSON_SCHEMA });
}
function schemaModuleOpts2File(schemaFile, vaultPath, fname) {
    const ext = ".schema.yml";
    return fs_extra_1.default.writeFile(path_1.default.join(vaultPath, fname + ext), serializeModuleOpts(schemaFile));
}
exports.schemaModuleOpts2File = schemaModuleOpts2File;
function schemaModuleProps2File(schemaMProps, vpath, fname) {
    const ext = ".schema.yml";
    return fs_extra_1.default.writeFile(path_1.default.join(vpath, fname + ext), common_all_1.SchemaUtils.serializeModuleProps(schemaMProps));
}
exports.schemaModuleProps2File = schemaModuleProps2File;
function assignJSONWithComment(jsonObj, dataToAdd) {
    return (0, comment_json_1.assign)({
        ...dataToAdd,
    }, jsonObj);
}
exports.assignJSONWithComment = assignJSONWithComment;
async function readJSONWithComments(fpath) {
    const content = await fs_extra_1.default.readFile(fpath);
    const obj = (0, comment_json_1.parse)(content.toString());
    return obj;
}
exports.readJSONWithComments = readJSONWithComments;
function readJSONWithCommentsSync(fpath) {
    const content = fs_extra_1.default.readFileSync(fpath);
    const obj = (0, comment_json_1.parse)(content.toString());
    return obj;
}
exports.readJSONWithCommentsSync = readJSONWithCommentsSync;
function tmpDir() {
    const dirPath = (0, tmp_1.dirSync)();
    return dirPath;
}
exports.tmpDir = tmpDir;
/** Returns the path to where the notes are stored inside the vault.
 *
 * For self contained vaults, this is the `notes` folder inside of the vault.
 * For other vault types, this is the root of the vault itself.
 *
 * If you always need the root of the vault, use {@link pathForVaultRoot} instead.
 */
const vault2Path = ({ vault, wsRoot, }) => {
    return (0, files_1.resolvePath)(common_all_1.VaultUtils.getRelPath(vault), wsRoot);
};
exports.vault2Path = vault2Path;
/** Returns the root of the vault.
 *
 * This is similar to {@link vault2Path}, the only difference is that for self
 * contained vaults `vault2Path` returns the `notes` folder inside the vault,
 * while this returns the root of the vault.
 */
function pathForVaultRoot({ vault, wsRoot, }) {
    if (common_all_1.VaultUtils.isSelfContained(vault))
        return (0, files_1.resolvePath)(path_1.default.join(wsRoot, vault.fsPath));
    return (0, exports.vault2Path)({ vault, wsRoot });
}
exports.pathForVaultRoot = pathForVaultRoot;
function writeJSONWithCommentsSync(fpath, data) {
    const payload = (0, comment_json_1.stringify)(data, null, 4);
    return fs_extra_1.default.writeFileSync(fpath, payload);
}
exports.writeJSONWithCommentsSync = writeJSONWithCommentsSync;
async function writeJSONWithComments(fpath, data) {
    const payload = (0, comment_json_1.stringify)(data, null, 4);
    return fs_extra_1.default.writeFile(fpath, payload);
}
exports.writeJSONWithComments = writeJSONWithComments;
/**
 * Turn . delimited file to / separated
 */
function dot2Slash(fname) {
    return fname.replace(/\./g, "/");
}
exports.dot2Slash = dot2Slash;
/** Checks that the `path` contains a file. */
async function fileExists(path) {
    try {
        const stat = await fs_extra_1.default.stat(path);
        return stat.isFile();
    }
    catch {
        return false;
    }
}
exports.fileExists = fileExists;
/** Finds if a file `fpath` is located in any vault.
 *
 * @param fpath A file name or relative path that we are searching inside vaults.
 */
async function findFileInVault({ fpath, wsRoot, vaults, }) {
    // Assets from later vaults will overwrite earlier ones.
    vaults = [...vaults].reverse();
    for (const vault of vaults) {
        const fullPath = path_1.default.join(wsRoot, common_all_1.VaultUtils.getRelPath(vault), fpath);
        // Doing this sequentially to simulate how publishing handles conflicting assets.
        // eslint-disable-next-line no-await-in-loop
        if (await fileExists(fullPath)) {
            return { vault, fullPath };
        }
    }
    return;
}
async function findNonNoteFile(opts) {
    let { fpath } = opts;
    if (path_1.default.isAbsolute(fpath)) {
        // The path could be an absolute path. If it is and the file exists, then directly use that.
        if (await fileExists(fpath))
            return { fullPath: fpath };
    }
    // Not an absolute path. Then the leading slash is meaningless:
    // `/assets` and `assets` refers to the same place.
    fpath = lodash_1.default.trim(fpath, "/\\");
    // Check if this is an asset first
    if (fpath.startsWith("assets")) {
        const out = await findFileInVault(opts);
        if (out !== undefined)
            return out;
    }
    // If not an asset, this also might be relative to the current note
    if (opts.currentVault) {
        const fullPath = path_1.default.join(opts.wsRoot, common_all_1.VaultUtils.getRelPath(opts.currentVault), fpath);
        if (await fileExists(fullPath))
            return { fullPath, vault: opts.currentVault };
    }
    // If not an asset, or if we couldn't find it in assets, then check from wsRoot for out-of-vault files
    const fullPath = path_1.default.join(opts.wsRoot, fpath);
    if (await fileExists(fullPath))
        return { fullPath };
    // Otherwise, it just doesn't exist
    return undefined;
}
exports.findNonNoteFile = findNonNoteFile;
class FileUtils {
    /**
     * Keep incrementing a numerical suffix until we find a path name that does not correspond to an existing file
     * @param param0
     */
    static genFilePathWithSuffixThatDoesNotExist({ fpath, sep = "-", }) {
        // Try to put into `fpath`. If `fpath` exists, create a new folder with an numbered suffix
        let acc = 0;
        let tryPath = fpath;
        while (fs_extra_1.default.pathExistsSync(tryPath)) {
            acc += 1;
            tryPath = [fpath, acc].join(sep);
        }
        return { filePath: tryPath, acc };
    }
}
exports.FileUtils = FileUtils;
_a = FileUtils;
/**
 * Check if a file starts with a prefix string
 * @param fpath: full path to the file
 * @param prefix: string prefix to check for
 */
FileUtils.matchFilePrefix = async ({ fpath, prefix, }) => {
    // solution adapted from https://stackoverflow.com/questions/70707646/reading-part-of-file-in-node
    return new Promise((resolve) => {
        const fileStream = fs_extra_1.default.createReadStream(fpath, { highWaterMark: 60 });
        const prefixLength = prefix.length;
        fileStream
            .on("error", (err) => resolve({
            error: new common_all_1.DendronError({ innerError: err, message: "error" }),
        }))
            // we got to the end without a match
            .on("end", () => resolve({ data: false }))
            .on("data", (chunk) => {
            // eslint-disable-next-line no-plusplus
            for (let i = 0; i < chunk.length; i++) {
                const a = String.fromCharCode(chunk[i]);
                // not a match, return
                if (a !== prefix[i]) {
                    resolve({ data: false });
                }
                // all matches
                if (i === prefixLength - 1) {
                    resolve({ data: true });
                }
            }
        });
    });
};
/** Looks at the files at the given path to check if it's a self contained vault. */
async function isSelfContainedVaultFolder(dir) {
    return lodash_1.default.every(await Promise.all([
        fs_extra_1.default.pathExists(path_1.default.join(dir, common_all_1.CONSTANTS.DENDRON_CONFIG_FILE)),
        fs_extra_1.default.pathExists(path_1.default.join(dir, common_all_1.FOLDERS.NOTES)),
    ]));
}
exports.isSelfContainedVaultFolder = isSelfContainedVaultFolder;
/** Move a file or folder from `from` to `to`, if the file exists.
 *
 * @returns True if the file did exist and was moved successfully, false otherwise.
 */
async function moveIfExists(from, to) {
    try {
        if (await fs_extra_1.default.pathExists(from)) {
            await fs_extra_1.default.move(from, to);
            return true;
        }
    }
    catch (err) {
        // Permissions error or similar issue when moving the path
        // deliberately left empty
    }
    return false;
}
exports.moveIfExists = moveIfExists;
/** Utility functions for dealing with file extensions. */
class FileExtensionUtils {
    static ensureTextExtensions() {
        if (this.textExtensions === undefined) {
            this.textExtensions = new Set(textextensions_1.default.map((extension) => extension.toLowerCase()));
        }
    }
    /** Checks if a given file extension is a well known text file extension. */
    static isTextFileExtension(extension) {
        extension = lodash_1.default.trimStart(extension, ".").toLowerCase();
        this.ensureTextExtensions();
        return this.textExtensions.has(extension);
    }
}
exports.FileExtensionUtils = FileExtensionUtils;
