// @dendronhq/common-server/lib/StateValidator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateValidator = void 0;
const common_all_1 = require("@dendronhq/common-all");
const path_1 = __importDefault(require("path"));
const files_1 = require("./files");
const filesv2_1 = require("./filesv2");
/**
 * Validate correctness of engine notes compared with filesystem.
 * Currently supports parent/children relationship between notes
 */
class StateValidator {
    static validateEngineState(engine) {
        return Promise.all(engine.vaults.map(async (vault) => {
            const rootNote = (await engine.findNotes({ fname: "root", vault }))[0];
            try {
                if (rootNote) {
                    const engineTreeNode = common_all_1.TreeUtils.createTreeFromEngine(engine.notes, rootNote.id);
                    const vpath = (0, filesv2_1.vault2Path)({ vault, wsRoot: engine.wsRoot });
                    const out = await (0, files_1.getAllFiles)({
                        root: common_all_1.URI.file(vpath),
                        include: ["*.md"],
                    });
                    if (out.error) {
                        return { error: out.error };
                    }
                    if (out.data) {
                        const allFnames = out.data.map((filePath) => path_1.default.parse(filePath).name);
                        const fileTreeNode = common_all_1.TreeUtils.createTreeFromFileNames(allFnames, "root");
                        const resp = common_all_1.TreeUtils.validateTreeNodes(fileTreeNode, engineTreeNode);
                        if (resp.error) {
                            return {
                                error: new common_all_1.DendronError({
                                    message: `Vault "${common_all_1.VaultUtils.getName(vault)}" mismatch. ${resp.error.message}`,
                                }),
                            };
                        }
                        return { data: undefined };
                    }
                    else {
                        return {
                            error: new common_all_1.DendronError({
                                message: `No files found from vault "${common_all_1.VaultUtils.getName(vault)}"`,
                            }),
                        };
                    }
                }
                else {
                    return {
                        error: new common_all_1.DendronError({
                            message: `Root file from vault "${common_all_1.VaultUtils.getName(vault)}" is missing.`,
                        }),
                    };
                }
            }
            catch (error) {
                const dendronError = common_all_1.ErrorFactory.wrapIfNeeded(error);
                return {
                    error: dendronError,
                };
            }
        }));
    }
}
exports.StateValidator = StateValidator;
