// @dendronhq/common-all/lib/types/configs/dendronConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genDefaultDendronConfig = void 0;
const commands_1 = require("./commands/commands");
const workspace_1 = require("./workspace/workspace");
const preview_1 = require("./preview/preview");
const publishing_1 = require("./publishing/publishing");
const global_1 = require("./global/global");
const dev_1 = require("./dev/dev");
/**
 * Generates a default DendronConfig using
 * respective default config generators of each sub config groups.
 * @returns DendronConfig
 */
function genDefaultDendronConfig() {
    return {
        global: (0, global_1.genDefaultGlobalConfig)(),
        commands: (0, commands_1.genDefaultCommandConfig)(),
        workspace: (0, workspace_1.genDefaultWorkspaceConfig)(),
        preview: (0, preview_1.genDefaultPreviewConfig)(),
        publishing: (0, publishing_1.genDefaultPublishingConfig)(),
        dev: (0, dev_1.genDefaultDevConfig)(),
    };
}
exports.genDefaultDendronConfig = genDefaultDendronConfig;
